[#11] 프로필 정보 수정

cyj199637 opened the pull request:
- 파일 관련 로직을 다루는 FileService 생성
- 로컬에서 파일을 다루는 LocalFileService 생성
- LocalFileService 파일 업로드, 파일 삭제 기능 구현
- UserServiceImpl - updateUser에서 프로필 정보까지 수정되도록 변경

cyj199637 committed:
[#11] 프로필 정보 수정 기능 추가

- 파일 관련 로직을 다루는 FileService 생성
- 로컬에서 파일을 다루는 LocalFileService 생성
- LocalFileService 파일 업로드, 파일 삭제 기능 구현
- UserServiceImpl - updateUser에서 프로필 정보까지 수정되도록 변경

cyj199637 commented:
UserController, UserServiceImpl의 변경된 로직에 관한 테스트 메소드 수정은 추후에 다시 마무리하겠습니다~!

f-lab-dev commented on src/main/java/me/liiot/snsserver/exception/InvalidValueException.java:
```diff
@@ -0,0 +1,12 @@
+package me.liiot.snsserver.exception;
+
+public class InvalidValueException extends Exception {
```
이거는 checked exception으로 만드신 이유가 있으신가요?
**잘못된 값을 입력했다**라는 것은 checked exception과 맞지 않는 것 같아서요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -0,0 +1,45 @@
+@Service
+public class LocalFileService implements FileService {
+
+    private static final String BASE_DIR = "C:\\Users\\cyj19\\Desktop\\Project\\sns-server\\images";
```
이거는 서버에 배포했을 때도 고려해서 외부에서 주입해주면 어떨까요?
`@Value` 어노테이션을 사용해볼 수 있을 것 같네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.UUID;
+
+@Service
+public class LocalFileService implements FileService {
```
업로드는 있는데 조회는 없네요~

cyj199637 replied:
조회 기능은 나중에 포스트 관련 기능을 만들 때 만들어도 될까요?
현재 파일이 프로필 이미지 밖에 없는데 이게 User의 한 필드에 묶여있다보니 프로필 이미지만의 조회 기능은 필요없다고 생각되어서요!

f-lab-dev replied:
네 그러면 그러셔도 됩니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -0,0 +1,45 @@
+
+        try {
+            targetFile.transferTo(new File(String.valueOf(filePathBuilder)));
+            FileInfo fileInfo = new FileInfo(newFileName, String.valueOf(filePathBuilder));
```
파일 경로도 DB에 저장해놔야하지 않을까요?

f-lab-dev commented on src/test/java/me/liiot/snsserver/service/UserServiceTest.java:
```diff
@@ -166,7 +182,7 @@ public void deleteUserTestWithSuccess() {
     @Test
     public void deleteUserTestWithFail() {
```
이 클래스에서 `FileUploadService`를 의존하는데 mocking하지 않은 이유가 있으신가요??

cyj199637 committed:
[#11] 피드백 반영

- local.file.base.directory 값을 외부에서 주입
- 수정된 로직에 따른 UserControllerTest, UserServiceTest 변경

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -2,17 +2,21 @@
 public class LocalFileService implements FileService {
 
-    private static final String BASE_DIR = "C:\\Users\\cyj19\\Desktop\\Project\\sns-server\\images";
+    @Value("${local.file.base.directory}")
```
이 property도 설계하실때 구조적으로 잘 설계해야하는데요~
itda 서비스에 종속적인 property이므로 `itda.~~` 로 설계해보시면 어떨까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -2,17 +2,21 @@
 
-    private static final String BASE_DIR = "C:\\Users\\cyj19\\Desktop\\Project\\sns-server\\images";
+    @Value("${local.file.base.directory}")
+    private String BASE_DIR;
```
이거는 이제 전역상수가 아니기때문에 카멜케이스를 지켜주세요~

cyj199637 committed:
[#11] AWS S3를 활용한 FileService 확장

- AwsFileService 추가
- 개발환경, 운영환경에 따른 profile 분리

cyj199637 committed:
[#11] 피드백 반영 및 FileService 수정

- 파일 저장 경로 수정
- FileService deleteDirectory 추가
- 회원정보 업데이트 성공시, 갱신된 User 객체가 세션에 저장되도록 수정

cyj199637 committed:
[#11] AwsFileService deleteDirectory 수정

cyj199637 commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -77,7 +77,7 @@ public void deleteFile(String filePath) {
-                .key("test1")
+                .key(userId)
                 .build());
     }
```
s3에서 객체를 삭제할 때 폴더 객체를 한 번에 삭제하는 것은 불가능한가요?

f-lab-dev replied:
이런 것들은 s3 document를 읽어보시면 좋을 것 같습니다~

f-lab-dev commented on pom.xml:
```diff
@@ -114,6 +114,26 @@
+        -->
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
```
이게 꼭 필요할까요? spring에 들어있지 않나요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -0,0 +1,66 @@
+public class AwsFileService implements FileService {
+
+    @Value("${aws.s3.bucket.name}")
+    String bucket;
```
접근제한자는 설정해주세요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -0,0 +1,66 @@
+
+            return fileInfo;
+        } catch (SdkServiceException | SdkClientException | IOException e) {
+            throw new FileUploadException("파일 업로드에 실패했습니다.");
```
여기에 발생한 exception 객체도 같이 cause에 넣어주면 추후 디버깅할때 좋지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FileService.java:
```diff
@@ -2,11 +2,26 @@
 
     void deleteFile(String filePath);
+
+    default String changeFileName(MultipartFile file) {
```
default 메소드를 활용하신게 흥미로운데요. 의도를 주석으로 남겨주시면 더 좋을 것 같습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/LocalFileService.java:
```diff
@@ -3,31 +3,31 @@
+        String newFileName = changeFileName(targetFile);
+
+        StringBuilder filePath = new StringBuilder();
+        filePath.append(baseDir).append("\\").append(newFileName);
```
`\\`가 리눅스에서는 `/`가 될 수도 있습니다. 어떻게 하는게 좋을까요?

cyj199637 committed:
[#11] 피드백 반영

- AwsFileService의 deleteDirectory() 메소드 수정
- file path saparator 수정
- exception 처리 수정

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/FileService.java:
```diff
@@ -17,12 +15,17 @@
+    /*
+    FileService 인터페이스를 구현하는 모든 클래스에서 중복 파일명, 한글 파일명과 같은 문제로
+    changeFileName() 메소드를 사용하고 있습니다. 그러나 각 클래스마다 changeFileName 로직이
+    같아 중복되는 코드가 발생했기 때문에 이를 제거하고자 인터페이스에서 default 메소드로 정의했습니다.
```
이런 의도라면 꼭 default 메소드로 적용해야할까요? 별도의 클래스로 분리하는건 어떨까요?

cyj199637 committed:
[#11] 피드백 반영

- changeFileName 별도의 클래스로 분리
